# Loan Default Prediction

<img src="http://jihoon-kim.synology.me/wp-content/uploads/2017/07/LOAN-1024x512.jpg">

## Import Modules

In [2]:
# import modules
import pandas as pd

# Load Data

Data is provided by Kaggle Datasets(https://www.kaggle.com/wendykan/lending-club-loan-data).

### Data Description
>These files contain complete loan data for all loans issued through the 2007-2015, including the current loan status (Current, Late, Fully Paid, etc.) and latest payment information. The file containing loan data through the "present" contains complete loan data for all loans issued through the previous completed calendar quarter. Additional features include credit scores, number of finance inquiries, address including zip codes, and state, and collections among others. The file is a matrix of about 890 thousand observations and 75 variables. A data dictionary is provided in a separate file.

In [4]:
# load data
loan = pd.read_csv('./data/loan.csv')

/home/jihoon_kim/anaconda3/envs/tensorflow/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2698: DtypeWarning: Columns (19,55) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


# Exploratory Data Analysis

Let's quickly explore how this data looks like.

In [7]:
loan.dtypes

id                               int64
member_id                        int64
loan_amnt                      float64
funded_amnt                    float64
funded_amnt_inv                float64
term                            object
int_rate                       float64
installment                    float64
grade                           object
sub_grade                       object
emp_title                       object
emp_length                      object
home_ownership                  object
annual_inc                     float64
verification_status             object
issue_d                         object
loan_status                     object
pymnt_plan                      object
url                             object
desc                            object
purpose                         object
title                           object
zip_code                        object
addr_state                      object
dti                            float64
delinq_2yrs              

In [8]:
loan.isnull().sum()

id                                  0
member_id                           0
loan_amnt                           0
funded_amnt                         0
funded_amnt_inv                     0
term                                0
int_rate                            0
installment                         0
grade                               0
sub_grade                           0
emp_title                       51457
emp_length                          0
home_ownership                      0
annual_inc                          4
verification_status                 0
issue_d                             0
loan_status                         0
pymnt_plan                          0
url                                 0
desc                           761350
purpose                             0
title                             151
zip_code                            0
addr_state                          0
dti                                 0
delinq_2yrs                        29
earliest_cr_

There are some columns with huge number of null values! I will drop these columns.

In [9]:
null_cols = ['desc',
             'mths_since_last_delinq',
             'mths_since_last_record',
             'next_pymnt_d',
             'mths_since_last_major_derog',
             'annual_inc_joint',
             'dti_joint',
             'verification_status_joint',
             'open_acc_6m',
             'open_il_6m',
             'open_il_12m',
             'open_il_24m',
             'mths_since_rcnt_il',
             'total_bal_il',
             'il_util',
             'open_rv_12m',
             'open_rv_24m',
             'max_bal_bc',
             'all_util',
             'inq_fi',
             'total_cu_tl',
             'inq_last_12m'
             ]

In [10]:
def drop_null_columns(data):
    """Drop columns (most of values are null)"""
    data.drop(null_cols, axis=1, inplace=True)
    return None

In [11]:
drop_null_columns(loan)

# Modifying Target Column

The target column I am interested in is `loan_status`. Let's explore how this column is composed of.

In [12]:
loan.loan_status.value_counts()

Current                                                601779
Fully Paid                                             207723
Charged Off                                             45248
Late (31-120 days)                                      11591
Issued                                                   8460
In Grace Period                                          6253
Late (16-30 days)                                        2357
Does not meet the credit policy. Status:Fully Paid       1988
Default                                                  1219
Does not meet the credit policy. Status:Charged Off       761
Name: loan_status, dtype: int64

Meaning of loan status is described below:

|Loan Status|Meaning|
|---:|:---|
|Current|Loan is up to date on all outstanding payments.|
|Fully Paid|Loan has been fully repaid, either at the expiration of the 3- or 5-year year term or as a result of a prepayment.|
|Charged Off|Loan for which there is no longer a reasonable expectation of further payments. Generally, Charge Off occurs no later than 30 days after the Default status is reached. Upon Charge Off, the remaining principal balance of the Note is deducted from the account balance.|
|Late (31-120 days)|Loan has not been current for 31 to 120 days.|
|Issued|New loan that has passed all LendingClub reviews, received full funding, and has been issued.|
|In Grace Period|Loan is past due but within the 15-day grace period.|
|Late (16-30 days)|Loan has not been current for 16 to 30 days.|
|Default|Loan has not been current for 121 days or more.|

First of all, `Current` and `Issued` loans are still in progress, so they cannot be used for training. Let's split them.

In [15]:
in_progress_index = ['Current', 'Issued']

In [16]:
def split_loan_in_progress(data):
    """Return table of loan in progress. It drops the loan in progress from loan data internally."""
    progress_bool = data.loan_status.isin(in_progress_index)
    loan_in_progress = data[progress_bool]
    data.drop(list(loan_in_progress.index), axis=0, inplace=True)
    return loan_in_progress

In [17]:
loan_in_progress = split_loan_in_progress(loan)

In [19]:
loan_in_progress

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
4        1075358    1311748     3000.0       3000.0      3000.000000   
6        1069639    1304742     7000.0       7000.0      7000.000000   
32       1065420    1299514    10000.0      10000.0      9975.000000   
39       1069346    1304237    12500.0      12500.0     12475.000000   
61       1068934    1277395    17500.0       8950.0      8925.000000   
86       1063958    1297940    14000.0      14000.0     13975.000000   
95       1068575    1303001    15300.0      15300.0     15275.000000   
99       1067874    1302235     6000.0       6000.0      6000.000000   
101      1034693    1264291    16000.0      16000.0     16000.000000   
167      1067654    1302043    16000.0      16000.0     15975.000000   
184      1046969    1277832    11000.0      11000.0     10750.000000   
187      1060875    1292842    22475.0      22475.0     22425.000000   
192      1067266    1301634     8400.0       8400.0      8150.000000   
217      1066664    1300858    21000.0      11800.0     11500.000000   
218      1066659    1300852    16000.0      16000.0     16000.000000   
239      1066344    1300716    15600.0      15600.0     15600.000000   
242      1064063    1296651    18825.0      18825.0     18800.000000   
261      1066173    1300526    12300.0      12300.0     12300.000000   
285      1065145    1299228    18000.0      18000.0     17975.000000   
321      1065342    1299434    20000.0      20000.0     19950.000000   
326      1065142    1299224    25000.0      25000.0     24975.000000   
327      1065320    1299408    18000.0      18000.0     18000.000000   
333       977277    1200058    20000.0      13475.0     13041.470438   
335      1059734    1291550    20000.0      20000.0     18626.346744   
338      1065104    1299181    20000.0      20000.0     18847.373057   
347      1064958    1299015    16000.0      11625.0     11368.734456   
361      1064897    1298948     3000.0       3000.0      3000.000000   
367      1064862    1298709    17000.0      17000.0     16060.110691   
375      1064639    1298460    25000.0      25000.0     23176.933614   
389      1064589    1298401     4125.0       4125.0      4125.000000   
...          ...        ...        ...          ...              ...   
887341  36421722   39143147    20000.0      20000.0     20000.000000   
887342  35034116   37717438    16000.0      16000.0     16000.000000   
887343  36870234   39622961     5000.0       5000.0      5000.000000   
887344  34303717   36967090    21000.0      21000.0     21000.000000   
887348  36631385   39374138    13000.0      13000.0     13000.000000   
887349  36691397   39434133    20000.0      20000.0     20000.000000   
887350  36500819   39232937    10300.0      10300.0     10300.000000   
887352  36231083   38942474    15000.0      15000.0     15000.000000   
887353  36371802   39103231    15000.0      15000.0     15000.000000   
887354  36250801   38962184     6000.0       6000.0      6000.000000   
887355  36331554   39042967    26950.0      26950.0     26950.000000   
887356  36460697   39182359    23000.0      23000.0     23000.000000   
887357  35773415   38468842    18700.0      18700.0     18700.000000   
887358  36440728   39152101    25000.0      25000.0     25000.000000   
887359  35591131   38288343    25000.0      25000.0     25000.000000   
887360  36361815   39093260    26500.0      26500.0     26500.000000   
887361  36201871   38913321    21000.0      21000.0     21000.000000   
887362  36381756   39013193     8000.0       8000.0      8000.000000   
887363  36231724   38943173    12000.0      12000.0     12000.000000   
887365  36381704   39013139     7000.0       7000.0      7000.000000   
887367  36301729   39023153    10000.0      10000.0     10000.000000   
887368  36250208   38961532    13150.0      13150.0     13150.000000   
887370  35094480   37777829     7500.0       7500.0      7500.000000   
887372  36330911   39042282    12000.0      12000.0   

In [20]:
loan

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0        1077501    1296599     5000.0       5000.0      4975.000000   
1        1077430    1314167     2500.0       2500.0      2500.000000   
2        1077175    1313524     2400.0       2400.0      2400.000000   
3        1076863    1277178    10000.0      10000.0     10000.000000   
5        1075269    1311441     5000.0       5000.0      5000.000000   
7        1072053    1288686     3000.0       3000.0      3000.000000   
8        1071795    1306957     5600.0       5600.0      5600.000000   
9        1071570    1306721     5375.0       5375.0      5350.000000   
10       1070078    1305201     6500.0       6500.0      6500.000000   
11       1069908    1305008    12000.0      12000.0     12000.000000   
12       1064687    1298717     9000.0       9000.0      9000.000000   
13       1069866    1304956     3000.0       3000.0      3000.000000   
14       1069057    1303503    10000.0      10000.0     10000.000000   
15       1069759    1304871     1000.0       1000.0      1000.000000   
16       1065775    1299699    10000.0      10000.0     10000.000000   
17       1069971    1304884     3600.0       3600.0      3600.000000   
18       1062474    1294539     6000.0       6000.0      6000.000000   
19       1069742    1304855     9200.0       9200.0      9200.000000   
20       1069740    1284848    20250.0      20250.0     19142.161077   
21       1039153    1269083    21000.0      21000.0     21000.000000   
22       1069710    1304821    10000.0      10000.0     10000.000000   
23       1069700    1304810    10000.0      10000.0     10000.000000   
24       1069559    1304634     6000.0       6000.0      6000.000000   
25       1069697    1273773    15000.0      15000.0     15000.000000   
26       1069800    1304679    15000.0      15000.0      8725.000000   
27       1069657    1304764     5000.0       5000.0      5000.000000   
28       1069799    1304678     4000.0       4000.0      4000.000000   
29       1047704    1278806     8500.0       8500.0      8500.000000   
30       1032111    1261745     4375.0       4375.0      4375.000000   
31       1069539    1304608    31825.0      31825.0     31825.000000   
...          ...        ...        ...          ...              ...   
887198  37790415   40553149    30000.0      30000.0     30000.000000   
887202  36118265   38829794    12000.0      12000.0     12000.000000   
887215  36400285   39031612    12825.0      12825.0     12825.000000   
887231  37830269   40593005     7675.0       7675.0      7675.000000   
887248  37117178   39889901    12000.0      12000.0     12000.000000   
887249  37357154   40119871    27650.0      27650.0     27650.000000   
887254  37068125   39536839     7200.0       7200.0      7200.000000   
887258  37147964   39920782    28500.0      28500.0     28500.000000   
887263  37317965   40080791    25000.0      25000.0     25000.000000   
887264  36801355   39554067    20000.0      20000.0     20000.000000   
887268  37297854   40040662    20000.0      20000.0     20000.000000   
887275  37247819   40020610    29675.0      29675.0     29675.000000   
887276  37257620   40030388    25000.0      25000.0     25000.000000   
887279  37127712   39900482     5875.0       5875.0      5875.000000   
887280  35998858   38700420     5875.0       5875.0      5875.000000   
887290  37167534   39940303    35000.0      35000.0     35000.000000   
887303  37087435   39850186    12000.0      12000.0     12000.000000   
887308  37227443   40000189     2725.0       2725.0      2725.000000   
887322  37317288   40080027    10000.0      10000.0     10000.000000   
887330  37077186   39839905    14000.0      14000.0     14000.000000   
887332  37187152   39959868    15000.0      15000.0     15000.000000   
887337  36200493   38911833    19600.0      19600.0     19600.000000   
887345  37011266   39773999    28000.0      28000.0     28000.000000   
887346  36808246   39560970     6000.0       6000.0   